[geos] Relocatable geos-config; dynamic builds on Unix

dg0yt opened the pull request:
- #### What does your PR fix?  
  This PR
  - makes `geos-config` relocatable, and adds a `geos-config-debug` for the debug build,
    as a step towards solving the gdal port issues (#9068, #8943)
  - fixes dynamic builds on Unix systems (#16602, #11681)

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  unchanged, no

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  yes

I still need to test the latest state with my gdal branch, but it is ready for vcpkg CI. The question in #17607 are related, too.

dg0yt committed:
Provide geos-config, and fix dynamic builds on Unix

dg0yt committed:
x-add-version

dg0yt committed:
Fix debug suffix handling

dg0yt committed:
Update git-tree

ras0219-msft commented on ports/geos/portfile.cmake:
```diff
@@ -34,19 +34,28 @@ vcpkg_configure_cmake(
+    string(REGEX REPLACE "(-lgeos(_c)?)d?( |\n)" "\\1d\\3" GEOS_CONFIG "${GEOS_CONFIG}")
+    file(WRITE "${CURRENT_PACKAGES_DIR}/debug/bin/geos-config" "${GEOS_CONFIG}")
+    file(MAKE_DIRECTORY ${CURRENT_PACKAGES_DIR}/share/${PORT})
+    file(RENAME ${CURRENT_PACKAGES_DIR}/debug/bin/geos-config ${CURRENT_PACKAGES_DIR}/share/${PORT}/geos-config-debug)
```
Can you give an example of how this would be consumed? My understanding is that `xyz-config` scripts like this are usually referred to like `"${XYZ_ROOT}/xyz-config"` and so having `xyz-config-debug` won't be directly usable. Instead, we should provide `debug/share/geos/geos-config` and set `GEOS_ROOT` to `debug/share/geos` when in debug mode.

dg0yt replied:
gdal `configure` says:
~~~
  --with-geos=ARG         Include GEOS support (ARG=yes, no or geos-config
                          path)
~~~
and similar for curl, kea, sfcgal.

So this is to be consumed in the gdal portfile by:
~~~
    list(APPEND OPTIONS_RELEASE --with-geos=${CURRENT_INSTALLED_DIR}/share/geos/geos-config)
    list(APPEND OPTIONS_DEBUG --with-geos=${CURRENT_INSTALLED_DIR}/share/geos/geos-config-debug)
~~~
I tested this locally, but this just one out of a pile of pending changes.

dg0yt replied:
>  provide `debug/share/geos/geos-config` and set set GEOS_ROOT to debug/share/geos

If I adopt this: Outside vcpkg, the `<package>_ROOT` variable generally (e.g. in CMake find modules) means the installation *prefix*, i.e. `${CURRENT_INSTALLED_DIR}`. 
As a consumer in a port file, I would still have to assign the roots to per-build-type parameters because the build types are handled deeper down in the cmake call stack.

dg0yt replied:
> provide `debug/share/geos/geos-config`

I tried that when preparing the PR, and I verified today: It leads into a build error:
~~~
-- Performing post-build validation
/debug/share should not exist. Please reorganize any important files, then use
    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/share")
~~~

JackBoosY replied:
Which file is in _${CURRENT_PACKAGES_DIR}/debug/share_?

dg0yt replied:
> Which file is in _${CURRENT_PACKAGES_DIR}/debug/share_?

None with the current PR. @ras0219-msft proposed to have `debug/share/geos/geos-config`, and I reported what would be the result.

dg0yt replied:
The alternative to fixing geos-config, curl-config etc. is forcing the consumers to use pkg-config where possible, by removing these files. Consumers also need to know whether to ask the scripts for static or standard flags, and this is solved for `PKG_CONFIG`.
E.g. port gdal will need a number of patches anyway, e.g. to use pkg-config for tiff.

JackBoosY commented:
Ping @ras0219-msft for review this PR again.

dg0yt commented:
More side notes: 
- This port sets the debug postfix ('d'). This might be contrary to the current maintainer guidelines. `git grep geosd` finds a small number of ports which depend on this.
- pc files do not carry the debug postfix.
- pc files are not passed to fixup. (Prefix pointing to packages.)
- Static linking dependencies are incomplete in the pc files.

dg0yt commented:
Will add pc file handling soon.

dg0yt committed:
Don't build benchmarks

dg0yt committed:
Revise pc file and geos-config handling

Fixup pc files.
Add -lgeos to Libs.private to satisfy common usage requirement
(already present in geos-config).
Install executable geos-config into tools.
Rewrite geos-config relocatability as patch.

dg0yt committed:
Port to non-deprecated cmake functions

dg0yt committed:
Add usage

dg0yt committed:
Update git-tree

dg0yt commented:
New changes:
- pc files rectified.
- geos-config moved to tools/bin and tools/debug/bin.
- Ported to vcpkg_configure_....
- Building of benchmarks disabled.
- Usage instructions added.

JackBoosY commented on ports/geos/portfile.cmake:
```diff
@@ -23,30 +25,45 @@ else()
+if(EXISTS "${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig/geos.pc")
+    geos_add_debug_postfix("${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig/geos.pc")
+endif()
+if(EXISTS "${CURRENT_PACKAGES_DIR}/bin/geos-config")
+    file(MAKE_DIRECTORY "${CURRENT_PACKAGES_DIR}/tools/${PORT}/bin")
+    file(RENAME "${CURRENT_PACKAGES_DIR}/bin/geos-config" "${CURRENT_PACKAGES_DIR}/tools/${PORT}/bin/geos-config")
+endif()
+if(EXISTS "${CURRENT_PACKAGES_DIR}/debug/bin/geos-config")
+    file(MAKE_DIRECTORY "${CURRENT_PACKAGES_DIR}/tools/${PORT}/debug/bin")
+    file(RENAME "${CURRENT_PACKAGES_DIR}/debug/bin/geos-config" "${CURRENT_PACKAGES_DIR}/tools/${PORT}/debug/bin/geos-config")
+    geos_add_debug_postfix("${CURRENT_PACKAGES_DIR}/tools/${PORT}/debug/bin/geos-config")
 endif()
```
Can we use `vcpkg_copy_tools` instead?

dg0yt replied:
> Can we use `vcpkg_copy_tools` instead?

No, no, and no.
- `vcpkg_copy_tools` doesn't handle tools without `VCPKG_TARGET_EXECUTABLE_SUFFIX`.
- `vcpkg_copy_tools` doesn't handle tools for the debug variant.
- For release+debug, the filesystem layout practice of `vcpkg_configure_make` is different from `vcpkg_copy_tools`.

File system layout for tools: That's why I wrote #17607, 23 days ago. Not to forget #13649. This is a blocker for quality of contributions, and also for vcpkg adoption.

strega-nil-ms commented:
Cool, LGTM! Thanks @dg0yt :)
